### PR TITLE
Selective traces

### PR DESCRIPTION
**Pull request by fishkerez**

### CODE REVIEW

File: plugins/taper/run.go [hunk 1]
```diff
@@ -79,6 +81,15 @@ func run(c *cli.Context) {
 		apiClient: apiClient,
 	}
 
+	if runConfig.EnableTraceSampling {
+		TSM, err := tsm.Create(false, runConfig.TraceSamplingManagerAddressEnv)
```
Comment (FrimIdan) [inline plugins/taper/run.go:85]: `false` ?

File: plugins/taper/run.go [hunk 2]
```diff
@@ -123,6 +134,10 @@ func (a *Agent) startReadOutputItems(ctx context.Context, outputItems chan *api.
 					log.Errorf("Failed to create telemetry: %v", err)
 					return
 				}
+				if a.traceSamplingManager != nil && !a.traceSamplingManager.ShouldTrace(telemetry.Request.Host) {
+					log.Infof("Ignoring host: %v", telemetry.Request.Host)
+					return
+				}
```
Comment (FrimIdan) [inline plugins/taper/run.go:140]: shouldn't it be part of `shouldIgnoreTelemetry`?

File: plugins/gateway/tyk/v3.2.2/plugin.go [hunk 3]
```diff
@@ -73,27 +88,38 @@ func PostGetAPIDefinition(_ http.ResponseWriter, r *http.Request) {
...
 }
 
 // Called during response phase.
 //nolint:deadcode
 func ResponseSendTelemetry(_ http.ResponseWriter, res *http.Response, req *http.Request) {
 	logger.Info("Handling telemetry")
+
+	if traceSamplingEnabled {
+		metadata := ctx.GetSession(req).MetaData
+		ignoreHost, ok := metadata[common.IgnoreHost].(string)
```
Comment (FrimIdan) [inline plugins/gateway/tyk/v3.2.2/plugin.go:113]: can't we run `TraceSamplingManager.ShouldTrace(host)` here and avoid writing/reading to session?

File: plugins/gateway/kong/plugin.go [hunk 4]
```diff
@@ -29,39 +29,71 @@ import (
...
 	_ = kong.Log.Info("Handling telemetry")
-	if conf.apiClient == nil {
+	if conf.TraceSamplingEnabled {
+		shouldIgnore, err := getShouldIgnoreHostFromContext(kong)
+		if err != nil {
+			_ = kong.Log.Err(fmt.Sprintf("Failed to get should ignore host from context: %v", err))
+		}
+		if shouldIgnore == "true" {
+			_ = kong.Log.Info("Ignoring host")
+			return
```
Comment (FrimIdan) [inline plugins/gateway/kong/plugin.go:81]: can't we run `TraceSamplingManager.ShouldTrace(host)` here and avoid writing/reading to context?

File: plugins/common/tsm/tsm.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+func (t *TraceSamplingManager) ShouldTrace(host string) bool {
+	hosts := t.getHosts()
+	if len(hosts) == 0 {
+		return true
+	}
+
+	for _, s := range hosts {
+		if s == host {
```
Comment (FrimIdan) [inline plugins/common/tsm/tsm.go:66]: need to address wildcard tracing.
if  s == "*" ---> return true

Comment (FrimIdan) [reply plugins/common/tsm/tsm.go:66]: https://github.com/apiclarity/wasm-filters/pull/8/files#diff-76cb46a2e63642865c78be467b0fd441981e1c449b5e847160c3a166c908f120R562

File: plugins/common/tsm/tsm.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+	"github.com/apiclarity/apiclarity/plugins/common"
+	"github.com/apiclarity/trace-sampling-manager/api/client/client"
+	"github.com/apiclarity/trace-sampling-manager/api/client/client/operations"
+)
+
+type TraceSamplingManager struct {
+	TraceSamplingManagerClient *client.TraceSamplingManager
+	// local cache of hosts to trace
+	Hosts            []string
```
Comment (FrimIdan) [inline plugins/common/tsm/tsm.go:18]: I think it is better to save it as a map so the lookup will be quicker. DYT?

File: backend/pkg/backend/backend.go [hunk 7]
```diff
@@ -103,14 +106,41 @@ func Run() {
...
+			log.Errorf("Failed to start trace sampling manager: %v", err)
+			return
+		}
+		// TODO remove
+		if config.HostToTraceDebug != "" {
+			tsm.SetHostsToTrace(&_interface.HostsToTrace{
+				Hosts:       []string{config.HostToTraceDebug},
+				ComponentID: "comp1",
+			})
+		}
```
[the remark this answers is not in the document]
Comment (fishkerez) [reply backend/pkg/backend/backend.go:143]: once I will finish testing

File: plugins/common/tsm/tsm.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		}
+	}
+	apiClient, err := common.NewTraceSamplingAPIClient(host, tlsOptions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new trace sampling api client: %v", err)
+	}
+
+	return &TraceSamplingManager{
+		TraceSamplingManagerClient: apiClient,
+		samplingInterval:           10 * time.Second,
```
[the remark this answers is not in the document]
Comment (fishkerez) [reply plugins/common/tsm/tsm.go:37]: Dont think it is crucial currently. you can offer a better interval if you think

File: plugins/gateway/tyk/v3.2.2/plugin.go (hunk 3 above)
Comment (fishkerez) [reply plugins/gateway/tyk/v3.2.2/plugin.go]: yes. thank you.

File: plugins/taper/run.go (hunk 1 above)
Comment (fishkerez) [reply plugins/taper/run.go:85]: dont enable tls

File: plugins/taper/run.go (hunk 2 above)
Comment (fishkerez) [reply plugins/taper/run.go:140]: Dont think so. shouldIgnoreTelemetry is a different logic of ignoring, it is more like filtering. 
Also, createTelemetry is extracting the host, I didnt want to extract it twice

File: plugins/common/tsm/tsm.go [hunk 9]
```diff
@@ -57,29 +60,23 @@ func (t *TraceSamplingManager) Start() {
...
+	return false
 }
 
 func (t *TraceSamplingManager) setHosts(hosts []string) {
 	t.lock.Lock()
 	defer t.lock.Unlock()
 
-	t.Hosts = hosts
+	for _, host := range hosts {
+		t.Hosts[host] = true
```
Comment (FrimIdan) [inline plugins/common/tsm/tsm.go:96]: need to create clear `t.Hosts` first

File: charts/apiclarity/values.yaml [hunk 10]
```diff
@@ -72,6 +72,9 @@ apiclarity:
       memory: "1000Mi"
       cpu: "200m"
 
+  traceSampling:
+    enable: false
```
Comment (FrimIdan) [inline charts/apiclarity/values.yaml]: shouldn't it be a global config?

File: plugins/taper/run.go (hunk 1 above)
Comment (FrimIdan) [reply plugins/taper/run.go:85]: shouldn't we get it from config or it is not supported for tap?

File: plugins/common/tsm/tsm.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (FrimIdan) [reply plugins/common/tsm/tsm.go:54]: I assume we will need it configurable eventually

File: backend/go.mod [hunk 11]
```diff
@@ -23,13 +24,16 @@ require (
...
-	k8s.io/api v0.21.3
-	k8s.io/apimachinery v0.21.3
-	k8s.io/client-go v0.21.3
+	k8s.io/api v0.23.5
+	k8s.io/apimachinery v0.23.5
+	k8s.io/client-go v0.23.5
 	sigs.k8s.io/yaml v1.2.0
 )
 
+replace github.com/apiclarity/trace-sampling-manager/api v0.0.0 => github.com/apiclarity/trace-sampling-manager/api v0.0.0-20220412134211-9c49bc797f35
```
Comment (akpsgit) [inline backend/go.mod:35]: don't forget to update to master after push

File: backend/pkg/backend/backend.go [hunk 12]
```diff
@@ -103,14 +106,43 @@ func Run() {
...
+			})
+			if err != nil {
+				log.Errorf("Failed to create a trace sampling manager: %v", err)
+				return
+			}
+			if err := tsm.Start(errChan); err != nil {
+				log.Errorf("Failed to start trace sampling manager: %v", err)
+				return
+			}
+			// TODO remove
```
Comment (akpsgit) [inline backend/pkg/backend/backend.go:138]: don't forget

File: backend/pkg/backend/backend.go [hunk 13]
```diff
@@ -45,6 +46,8 @@ import (
 	_spec "github.com/apiclarity/speculator/pkg/spec"
 	_speculator "github.com/apiclarity/speculator/pkg/speculator"
 	_mimeutils "github.com/apiclarity/speculator/pkg/utils"
+	"github.com/apiclarity/trace-sampling-manager/manager/pkg/manager"
+	_interface "github.com/apiclarity/trace-sampling-manager/manager/pkg/manager/interface"
```
Comment (akpsgit) [inline backend/pkg/backend/backend.go:50]: remove

File: backend/pkg/backend/backend.go [hunk 14]
```diff
@@ -103,14 +106,43 @@ func Run() {
 	dbHandler.StartReviewTableCleaner(globalCtx, time.Duration(config.DatabaseCleanerIntervalSec)*time.Second)
 
 	var monitor *k8smonitor.Monitor
+	var clientset kubernetes.Interface
+	var tsm *manager.Manager
```
Comment (akpsgit) [inline backend/pkg/backend/backend.go:110]: tsm -> samplingManager

File: backend/pkg/config/config.go [hunk 15]
```diff
@@ -26,12 +26,16 @@ import (
...
 
 const (
-	BackendRestPort            = "BACKEND_REST_PORT"
-	HTTPTracesPort             = "HTTP_TRACES_PORT"
-	HealthCheckAddress         = "HEALTH_CHECK_ADDRESS"
-	StateBackupIntervalSec     = "STATE_BACKUP_INTERVAL_SEC"
-	DatabaseCleanerIntervalSec = "DATABASE_CLEANER_INTERVAL_SEC"
-	StateBackupFileName        = "STATE_BACKUP_FILE_NAME"
+	BackendRestPort              = "BACKEND_REST_PORT"
+	HostToTraceDebug             = "HOST_TO_TRACE"
```
Comment (akpsgit) [inline backend/pkg/config/config.go:30]: to remove

File: plugins/common/tsm/tsm.go [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/apiclarity/apiclarity/plugins/common"
+	"github.com/apiclarity/trace-sampling-manager/api/client/client"
+	"github.com/apiclarity/trace-sampling-manager/api/client/client/operations"
+)
+
+type TraceSamplingManager struct {
```
Comment (akpsgit) [inline plugins/common/tsm/tsm.go:30]: TraceSamplingManager -> Client

File: plugins/common/tsm/tsm.go [hunk 17]
```diff
@@ -0,0 +1,107 @@
+// Copyright © 2021 Cisco Systems, Inc. and its affiliates.
```
[the remark this answers is not in the document]
Comment (akpsgit) [reply plugins/common/tsm/tsm.go:1]: trace-sampling-client/trace-sampling-client

File: plugins/common/tsm/tsm.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (akpsgit) [reply plugins/common/tsm/tsm.go:54]: let's move it as Create param

File: plugins/common/tsm/tsm.go [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+		}
+	}()
+}
+
+func (t *TraceSamplingManager) ShouldTrace(host string) bool {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	if len(t.Hosts) == 0 {
+		return true
```
Comment (akpsgit) [inline plugins/common/tsm/tsm.go:82]: return false

File: plugins/common/tsm/tsm.go [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+	}
+
+	return false
+}
+
+func (t *TraceSamplingManager) GetHosts() map[string]bool {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	ret := t.Hosts
```
Comment (akpsgit) [inline plugins/common/tsm/tsm.go:95]: return t.Hosts

File: plugins/common/tsm/tsm_test.go [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+					"*": true,
+				},
+			},
+			args:   args{
+				host: "host4",
+			},
+			want:   true,
+		},
+		{
+			name:   "should trace - empty list",
```
Comment (akpsgit) [inline plugins/common/tsm/tsm_test.go:142]: fix

File: plugins/common/tsm/tsm_test.go [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+		t1.Run(tt.name, func(t1 *testing.T) {
+			t := &TraceSamplingManager{
+				Hosts:                      tt.fields.Hosts,
+			}
+			if got := t.ShouldTrace(tt.args.host); got != tt.want {
+				t1.Errorf("ShouldTrace() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```
Comment (akpsgit) [inline plugins/common/tsm/tsm_test.go:162]: :)

File: plugins/gateway/kong/plugin.go [hunk 22]
```diff
@@ -29,39 +29,65 @@ import (
...
+	TraceSamplingEnabled bool   `json:"trace_sampling_enabled"`
 }
 
 func New() interface{} {
 	return &Config{}
 }
 
 func (conf Config) Access(kong *pdk.PDK) {
+	if conf.TraceSamplingEnabled && traceSamplingManager == nil {
+		_ = kong.Log.Info("Creating trace sampling manager")
```
Comment (akpsgit) [inline plugins/gateway/kong/plugin.go:51]: manager -> client

File: plugins/gateway/kong/plugin.go [hunk 23]
```diff
@@ -29,39 +29,65 @@ import (
...
+		traceSampling, err := tsm.Create(conf.EnableTLS, conf.TraceSamplingHost)
```
Comment (akpsgit) [inline plugins/gateway/kong/plugin.go:52]: traceSampling -> traceSamplingClient, pass interval here (from const)

File: plugins/gateway/kong/plugin.go [hunk 24]
```diff
@@ -29,39 +29,65 @@ import (
...
+		if err != nil {
+			_ = kong.Log.Err(fmt.Sprintf("Failed to create trace sampling manager: %v", err))
```
Comment (akpsgit) [inline plugins/gateway/kong/plugin.go:54]: manager -> client

File: plugins/gateway/tyk/v3.2.2/plugin.go [hunk 25]
```diff
@@ -47,18 +48,32 @@ const (
...
+	telemetryHost = os.Getenv("APICLARITY_HOST")
 	gatewayNamespace = os.Getenv("TYK_GATEWAY_NAMESPACE")
 	if os.Getenv("ENABLE_TLS") == "true" {
 		enableTLS = true
 	}
+	if os.Getenv("TRACE_SAMPLING_ENABLED") == "true" {
+		traceSamplingHost = os.Getenv("TRACE_SAMPLING_HOST_NAME")
+		traceSamplingManager, err := tsm.Create(enableTLS, traceSamplingHost)
+		if err != nil {
+			logger.Errorf("Failed to create trace sampling manager: %v", err)
```
Comment (akpsgit) [inline plugins/gateway/tyk/v3.2.2/plugin.go:70]: manager -> client

File: plugins/taper/config/config.go [hunk 26]
```diff
@@ -21,23 +21,29 @@ import (
...
 
 func LoadConfig() *Config {
 	return &Config{
-		NamespaceToTap:  viper.GetStringSlice(NamespacesToTapEnv),
-		UpstreamAddress: viper.GetString(UpstreamAddressEnv),
-		EnableTLS:       viper.GetBool(EnableTLSEnv),
+		NamespaceToTap:                 viper.GetStringSlice(NamespacesToTapEnv),
+		UpstreamAddress:                viper.GetString(UpstreamAddressEnv),
+		EnableTLS:                      viper.GetBool(EnableTLSEnv),
+		TraceSamplingManagerAddressEnv: viper.GetString(TraceSamplingManagerAddressEnv),
```
Comment (akpsgit) [inline plugins/taper/config/config.go:46]: TraceSamplingManagerAddress

File: plugins/taper/run.go [hunk 27]
```diff
@@ -79,6 +82,15 @@ func run(c *cli.Context) {
 		apiClient: apiClient,
 	}
 
+	if runConfig.TraceSamplingEnabled {
+		TSM, err := tsm.Create(false, runConfig.TraceSamplingManagerAddressEnv)
+		if err != nil {
+			log.Errorf("Failed to create trace sampling manager: %v", err)
```
Comment (akpsgit) [inline plugins/taper/run.go:88]: manager -> client

File: plugins/taper/run.go [hunk 28]
```diff
@@ -123,6 +135,10 @@ func (a *Agent) startReadOutputItems(ctx context.Context, outputItems chan *api.
 					log.Errorf("Failed to create telemetry: %v", err)
 					return
 				}
+				if a.traceSamplingManager != nil && !a.shouldTrace(telemetry.Request.Host) {
```
Comment (akpsgit) [inline plugins/taper/run.go:138]: a.traceSamplingManager != nil should be part of shouldtrace()

File: plugins/taper/run.go [hunk 29]
```diff
@@ -137,6 +153,26 @@ func (a *Agent) startReadOutputItems(ctx context.Context, outputItems chan *api.
...
 
+// since we dont have the destination namespace in tapper, we will compare host only by host name.
+func (a *Agent) shouldTrace(host string) bool {
+	hosts := a.traceSamplingManager.GetHosts()
+	if hosts["*"] {
+		// trace all
+		return true
+	}
+	if len(hosts) == 0 {
+		return true
```
Comment (akpsgit) [inline plugins/taper/run.go:164]: false

File: plugins/taper/run_test.go [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+			name:   "empty",
+			fields: fields{
+				traceSamplingManager: &tsm.TraceSamplingManager{
+					Hosts: map[string]bool{},
+				},
+			},
+			args:   args{
+				host: "host3",
+			},
+			want:   true,
```
Comment (akpsgit) [inline plugins/taper/run_test.go:78]: fix

File: plugins/taper/run.go [hunk 31]
```diff
@@ -137,6 +153,26 @@ func (a *Agent) startReadOutputItems(ctx context.Context, outputItems chan *api.
...
+	}
+	for h := range hosts {
+		// take only host name
+		h = strings.Split(h, ".")[0]
```
[the remark this answers is not in the document]
Comment (fishkerez) [reply plugins/taper/run.go:168]: will add a goroutine to do this in taper

[the remark this answers is not in the document]
Comment (fishkerez) [reply plugins/taper/run.go:168]: decided to match it with no changes

**Review by FrimIdan — APPROVED** [review]

**Pull request merged**